src/Propel/Generator/Command/Executor/RollbackExecutor.php

```diff
@@ -0,0 +1,196 @@
+<?php
+
+/**
+ * MIT License. This file is part of the Propel package.
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Propel\Generator\Command\Executor;
+
+use Exception;
+use Propel\Generator\Manager\MigrationManager;
+use Propel\Generator\Util\SqlParser;
+use Propel\Runtime\Exception\RuntimeException;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * Service class for executing rollback.
+ */
+class RollbackExecutor
+{
+    /**
+     * @var string
+     */
+    protected const COMMAND_OPTION_FAKE = 'fake';
+
+    /**
+     * @var string
+     */
+    protected const COMMAND_OPTION_FORCE = 'force';
+
+    /**
+     * @var string
+     */
+    protected const COMMAND_OPTION_VERBOSE = 'verbose';
+
+    /**
+     * @var \Symfony\Component\Console\Input\InputInterface
+     */
+    protected InputInterface $input;
+
+    /**
+     * @var \Symfony\Component\Console\Output\OutputInterface
+     */
+    protected OutputInterface $output;
+
+    /**
+     * @var \Propel\Generator\Manager\MigrationManager
+     */
+    protected MigrationManager $migrationManager;
+
+    /**
+     * @param \Symfony\Component\Console\Input\InputInterface $input
+     * @param \Symfony\Component\Console\Output\OutputInterface $output
+     * @param \Propel\Generator\Manager\MigrationManager $migrationManager
+     */
+    public function __construct(
+        InputInterface $input,
+        OutputInterface $output,
+        MigrationManager $migrationManager
+    ) {
+        $this->input = $input;
+        $this->output = $output;
+        $this->migrationManager = $migrationManager;
+    }
+
+    /**
+     * @param int $currentVersion
+     * @param int|null $previousVersion
+     *
+     * @return bool
+     */
+    public function executeRollbackToPreviousVersion(int $currentVersion, ?int $previousVersion = null): bool
+    {
+        $this->output->writeln(sprintf(
+            'Executing migration %s down',
+            $this->migrationManager->getMigrationClassName($currentVersion),
+        ));
+
+        $migration = $this->migrationManager->getMigrationObject($currentVersion);
+        if (!$this->isFake() && $migration->preDown($this->migrationManager) === false) {
+            if (!$this->isForce()) {
```

Comment:
* Could you rewrite the conditions to get rid of this level of nesting?

Reply:
@abitskil Can you give an example of what you have in mind?

Reply:
At a glance, I am not sure that would work, these are actually nested conditions. You can make it look different, like `if (a && b) {...} else if (a) {...}`, but I am  not sure if that is better. To me, there is no obvious "good" solution here.
Which is why I would like to know what you, @abitskil, had in mind. It seems like three people are now thinking about your comment.